speedup sphinx build with caching pip

smoench opened the pull request:
<!--

If your pull request fixes a BUG, use the oldest maintained branch that contains
the bug (see https://symfony.com/roadmap for the list of maintained branches).

If your pull request documents a NEW FEATURE, use the same Symfony branch where
the feature was introduced (and `master` for features of unreleased versions).

-->

Commit:
speedup sphinx build/linting with caching doctrees and pip cache

ricknox commented:
cc: @OskarStark

javiereguiluz commented:
@smoench this is really nice. Thanks for all your help to improve the performance of the doc linter.

wouterj commented:
Thanks a lot!

In the CMF docs, we also cached `$HOME/.cache/pip`, to speed up pip. Maybe that's an option here as well?

smoench commented:
Sure, I was also thinking about that

OskarStark reviewed (approved):
Nice Improvement ✌🏻

smoench commented on .github/workflows/ci.yaml:
```diff
@@ -29,12 +29,26 @@ jobs:
+                        ${{ runner.os }}-pip-
+
             -   name: "Install Sphinx"
                 run: pip install --user sphinx
```
@OskarStark Why is this needed? Because `_build/.requirements.txt` has also a reference for sphinx but in version 1.3.6

OskarStark replied:
Not sure, let’s try without...
Initially I used the same approach as on Travis

wouterj replied:
It wasn't part of the travis config: https://github.com/symfony/symfony-docs/pull/13061/files#diff-354f30a63fb0907d4ad57269548329e3

So I guess it has something to do with GitHub actions or otherwise, can be removed

OskarStark replied:
Hmm strange, maybe it was part of .symfony.cloud.yaml ?

smoench replied:
I removed it, works fine.

smoench commented:
I don't know what causes following error:
```
Warning, treated as error:
/home/runner/work/symfony-docs/symfony-docs/form/form_customization.rst:None: WARNING: duplicate label form-customization-sidebar, other instance in /home/runner/work/symfony-docs/symfony-docs/form/form_themes.rst
```

wouterj commented:
> I don't know what causes following error:

Hmm, that reference target (`form-customization-sidebar`) lives in `form/form_customization.rst` in 3.4 and in `form/form_themes.rst` for the newer versions. Seems like the cache is maybe shared among all branches, resulting in this error? Or does the docs somehow render a wrong version of the documentation?

smoench commented:
On the first try the base branch was on `master` but changed I changed it to `3.4`. I thought it wouldn't matter as of the cache restrictions https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache.
I changed cache keys now.

But some how after the cache hit and cache file has been extracted, sphinx thinks all files has been changed. So cache does not work properly 🤷‍♂

OskarStark commented:
So we only go with the pip cache and not the other ones?

smoench commented:
Unfortunately, yes! But perhaps someone can confirm that it is working with a more latest sphinx version than 1.3.6.

smoench commented on .github/workflows/ci.yaml:
```diff
@@ -29,12 +29,29 @@ jobs:
+                        ${{ runner.os }}-pip-
 
-            -   name: "Install custom requirements via pip"
+            -   name: "Install spinx + requirements via pip"
```
done

smoench commented:
Well, Sphinx uses `mtime` (in all versions) for detecting changes. This works on a local machine but not in a CI that makes a fresh checkout.

OskarStark commented:
Thank you Simon.
